Better handling in cases where zephyr_module is unable to get the git revision

fiveohhh opened the pull request:
I ran into an issue in my ci system that the git directory for my project git in an unexpected state.  The root cause ended up being some file permission issues with the .git dir.  It took quite awhile to track this all down though.  This MR makes it so that a more useful error message is bubbled up to the user, instead of a generic `TypeError: unsupported operand type(s) for +=: 'NoneType' and 'str'`

tejlmand reviewed (changes requested):
Thanks for this discovery, but Zephyr module would not be the right place to decide if this git issues should be treated as an error.

Zephyr module at this location is simply creating a meta object with information regarding the system, much like reporting a status. Of course returning some projects to be in unknown / broken state is acceptable, but it should not error out on the process here.

I would assume that other `west` command would fail, for example like `west update`.

Can you provide some more details regarding the state of your work space, and the result of commands like:
`west update`
`west status`

fiveohhh commented on scripts/zephyr_module.py:
```diff
@@ -630,6 +630,10 @@ def process_meta(zephyr_base, west_projs, modules, extra_modules=None,
                 off = True
 
             if off:
+                if meta_project['revision'] == None:
+                    # This would occur if there was an issue with retrieving the git
+                    # revision in _create_meta_project.git_revision(path)
+                    raise TypeError(f"Was unable to get git revision for {meta_project.get('path')}")
```
Ahh, I see. So a better approach would be if git rev returned something like `"unknown"` instead of `None` if it couldn't get the rev?

tejlmand replied:
Perhaps `unknown` is a better option, but before settling on a name, then it would be nice to know exactly what is reported by the git commands, see https://github.com/zephyrproject-rtos/zephyr/pull/89159#issuecomment-2838155223.

fiveohhh replied:
Sorry, that was a mistake in the original MR, it should have been before the `manifest_project["revision"] += "-off"` on line 610.  That is where the orignal error was happening. I was jumping between repos and edited the wrong spot.  But I think the other proposal that returns a string would help for both the issue I ran across, as well as the more general git issue.

tejlmand replied:
>  was jumping between repos and edited the wrong spot. 

I truly know the jumping back and fourth when tracing issues and their root cause :laughing: 

> I update this PR with the proposed "unknown" idea.

Thanks :+1:

fiveohhh commented:
> > Those commands are working, The way out setup works is we have our west project in mydir.
> 
> Interesting, because those commands also issues `rev-parse`, so the git command itself is working, but i'm curious of the value returned since the the following code fails.
> 
> Is it easy to create a failing project ? What does `git rev-parse --is-inside-work-tree` gives on a failing workspace ? What does `git rev-parse HEAD` gives on a failing workspace ?

They both give me the same `fatal: detected dubious ownership in repository `.   `west update` works just fine in this situation.  Does `west update` also update my project?  It looks like it is only doing external projects.  In this situation, it is only my project (`mydir/.git`) directory that has the incorrect file permissions.

fiveohhh commented:
> Thanks. Sounds likethe project referred to as `my project` is the [manifest repository](https://docs.zephyrproject.org/latest/develop/west/basics.html#workspace-concepts). 

Ahh, yes, that is correct.  Thanks for the references and correct terminology.  We're using the [T2 structure](https://docs.zephyrproject.org/latest/develop/west/workspaces.html#t2-star-topology-application-is-the-manifest-repository).

tejlmand commented on scripts/zephyr_module.py:
```diff
@@ -480,7 +480,7 @@ def git_revision(path):
                     return revision + '-dirty', True
                 return revision, False
-        return None, False
+        return "unknown", False
```
I think "unknown" is fine, but want to mention that this can happen in different scenarios, like:
- git returning error, like `fatal: detected dubious ownership`
- Folder is not a git repo

I think "unknown" is fine in both cases, but one could argue that there is a difference between a non-git repo, and a git repo which fails the git command, and thus different strings should be used to reflect this.

fiveohhh replied:
I could an argument either way.  Let me know if you have a preference and we can find a solution.

tejlmand replied:
As I wrote, then `unknown` is fine with me, and keeps the code simpler.

The reason for making the comment is, A) to make others aware that there are possibilities to differentiate, so they can comment if they have a preference. B) As a possible future reference in case topic is brought up again, then we know why this decision was made.

fiveohhh replied:
I see, that all makes sense.  Thanks for walking me through my first zephyr PR!

tejlmand reviewed (changes requested):
Thanks for the improved fix :+1: .

I believe we still need slight adjustments around here (and perhaps other places in addition):
https://github.com/zephyrproject-rtos/zephyr/blob/54a1a536f57a8440817e5ad00747f8f16cabc6e6/scripts/zephyr_module.py#L578-L579
https://github.com/zephyrproject-rtos/zephyr/blob/54a1a536f57a8440817e5ad00747f8f16cabc6e6/scripts/zephyr_module.py#L625

because `-off` is being appended to unknown, like this:
```
   projects:
   - path: <path>
     revision: unknown-off
```
which looks strange as the revision is already unknown.
So would prefer the result to be:
```
   projects:
   - path: <path>
     revision: unknown
```
which requires slight adjustment to the `-off` handling.

fiveohhh commented:
> I believe we still need slight adjustments around here (and perhaps other places in addition):

I added a check to each place revision is appended to, and it now only does the appending if revision is 40 char sha.

tejlmand reviewed (approved):
Thanks for the effort.

tejlmand commented:
@fiveohhh  you're missing a `signed-off-by` entry:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#dco-sign-off

See also the compliance failure: https://github.com/zephyrproject-rtos/zephyr/actions/runs/14764363042/job/41477548712?pr=89159
```
 9fc937ba5dc9d8ac4ac1970f0f724fe1cc8f0950: Missing signed-off-by line
Compliance error, check for error messages in the "Run Compliance Tests" step
You can run this step locally with the ./scripts/ci/check_compliance.py script.
Error: Process completed with exit code 1.
```

fiveohhh committed:
scripts: zephyr_module now  handles when unable to retrieve git rev

We now return "unknown" as the revision, and only concat onto
the revision if the revision is a valid git has value.

Signed-off-by: Andy Lee <andy@chiefmarley.com>

fiveohhh commented:
> @fiveohhh you're missing a signed-off-by entry

I just added it.  Thank you again for the help!
